### ISSUE TITLE
S3 upload fails when metadata is included

### ISSUE THREAD

**Issue by phantumcode** [awslabs/aws-sdk-swift#892]
### Describe the bug

S3 upload data fails with a 403 response when metadata is included in the request header

* Verified that S3 upload with the same request and metadata works in legacy SDK [AWSS3TrasnferUtility](https://github.com/aws-amplify/aws-sdk-ios/blob/main/AWSS3/AWSS3TransferUtility.m)
* Verified that permission is configured correctly in IAM and S3 PutObject is allowed for authorized, unauthorized, and guest users
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:DeleteObject"
            ],
            "Resource": [
                "arn:aws:s3:::xxxxxxxxx-xxx/public/*"
            ],
            "Effect": "Allow"
        }
    ]
}
```

### Expected Behavior

S3 upload data succeeds with 200 response when metadata is in the request header

### Current Behavior

S3 upload data fails with a http response of 403

### Reproduction Steps

Upload to S3 with a metdata in request header:
      
```swift
let input = PutObjectInput(bucket: "xxxxxxxxx-xxx", key: "public/test")
let preSignedUrl = try await input.presignURL(config: config, expiration: 3000)
var request = URLRequest(url: preSignedURL)
request.cachePolicy = .reloadIgnoringLocalCacheData
request.httpMethod = "PUT"
request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")

request.setValue("John Doe", forHTTPHeaderField: "x-amz-meta-author")

let uploadTask = urlSession.uploadTask(with: request, fromFile: fileURL)
uploadTask.resume()
```

### Additional Information/Context
Issue was initially submitted for Amplify Swift library https://github.com/aws-amplify/amplify-swift/issues/2744 
Amplify Swift library was ignoring/dropping metadata.  When metadata is added back, receiving 403 errors.

### AWS SWIFT SDK version used

0.6.1

### Compiler and Version used

Xcode 14.2, Swift 5.7

### Operating System and version

iOS 15.5

**Comment by phantumcode** [comment]
This appears to be an issue with the `presignURL` extension on `PutObjectInput`

Upload works as expected when uploading to s3 via S3Client:

```swift
            let s3client = try S3Client(config: ...)
            let input = PutObjectInput(
                body: dataStream,
                bucket: "xxxxxxxxx-xxx/public",
                key: "test",
                metadata: metadata
            )
            let response = try await s3client.putObject(input: input)
```

**Comment by phantumcode** [comment]
Also related/similar issue: upload via presigned url fails when key has special characters but succeeds when using S3Client directly: https://github.com/aws-amplify/amplify-swift/issues/2743

**Comment by ganeshnj** [comment]
This could be signature issue after adding the headers to the request later.

We need to investigate further. Essentially, the different between a presign and normal operation is not executing the request in presign operation. Otherwise, rest remain same.

**Comment by harsh62** [comment]
@epau  Would you be able to provide any updates on the issue?

**Comment by jbelkins** [comment]
@harsh62 @epau I'm looking at the SDK-side fix for this today: https://github.com/aws-amplify/amplify-swift/issues/2743

I'll look at this metadata issue next.  The fix is not the same for the two problems, but I suspect they are closely related in that they both involve proper handling of request components.

**Comment by jbelkins** [comment]
@harsh62 Metadata is normally sent in HTTP headers.  By default, HTTP headers are omitted (except the host header, which AWS requires to be in the signature) when generating a presigned URL.  This is why the metadata is not getting embedded in your presigned URL.

Additionally, you cannot set the metadata headers to be sent along with your presigned URL (like you tried above) because any header starting with `x-amz-` (which includes the metadata headers) has to be included in the code signature, and your request will fail with HTTP 403 / Forbidden.

The S3 docs do not document that metadata can be included in a query string (please link me the docs if you know where) but I can manually add metadata to a presigned URL's query string and it is accepted by S3.

So I can add this customization for S3 `PutObject` but this is a customization / feature, not a bug fix _per se_.  Let me know your priority for this and I can work to get it in a future release.

**Comment by phantumcode** [comment]
@jbelkins Thanks for the feedback.  I've updated my code sample to include the metadata as part of of getting the presigned url (see code example).  The S3 upload succeeds with no error.  When I check the S3 file in AWS Console, however, the metadata isn't there.
```swift
let metadata = ["author": "John Doe"]
let input = PutObjectInput(bucket: "xxxxxxxxx-xxx", key: "public/test", metadata: metadata)
let preSignedUrl = try await input.presignURL(config: config, expiration: 3000)
var request = URLRequest(url: preSignedURL)
request.cachePolicy = .reloadIgnoringLocalCacheData
request.httpMethod = "PUT"
let uploadTask = urlSession.uploadTask(with: request, fromFile: fileURL)
uploadTask.resume()
```

<img width="1569" alt="image" src="https://github.com/awslabs/aws-sdk-swift/assets/103537251/840cf6e3-7f4a-4fac-aa34-cbfcda9dd711">

**Comment by jbelkins** [comment]
@phantumcode Please see my comment above yours:

> So I can add this customization for S3 PutObject but this is a customization / feature, not a bug fix per se. Let me know your priority for this and I can work to get it in a future release.

I know presigned URLs don't currently include metadata.  It requires additional development work (i.e. customization) to add metadata to presigned URLs.  If Amplify wants metadata to be included on presigned URLs, please engage our team internally and we will determine a priority / estimate to get this feature added.

**Comment by jbelkins** [comment]
Released in AWS SDK for Swift 0.26.0
